File: Linked Lists/My Own Linked List/myOwnLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    # Add a new node to the end of the list
    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    # Insert a new node at a specific position
    def insert(self, data, position):
        new_node = Node(data)
        if position < 0:
            position = 0
        if position == 0:
            new_node.next = self.head
            self.head = new_node
            return
        current = self.head
        prev = None
        index = 0
        while current and index < position:
            prev = current
            current = current.next
            index += 1
        prev.next = new_node
        new_node.next = current

    # Remove a node at a specific position
    def remove_at(self, position):
        if not self.head or position < 0:
            return None
        if position == 0:
            removed = self.head
            self.head = self.head.next
            return removed
        current = self.head
        prev = None
        index = 0
        while current and index < position:
            prev = current
            current = current.next
            index += 1
        if current:
            prev.next = current.next
            return current
        return None

    # Find the index of a specific value
    def index_of(self, data):
        current = self.head
        index = 0
        while current:
            if current.data == data:
                return index
            current = current.next
            index += 1
        return -1

    # Check if the list is empty
    def is_empty(self):
        return not self.head

    # Get the size of the list
    def size(self):
        current = self.head
        count = 0
        while current:
            count += 1
            current = current.next
        return count

    # Convert the linked list to a Python list
    def to_list(self):
        result = []
        current = self.head
        while current:
            result.append(current.data)
            current = current.next
        return result


# Example usage:
linked_list = LinkedList()
linked_list.append(1)
linked_list.append(2)
linked_list.append(3)
linked_list.insert(0, 0)
removed_node = linked_list.remove_at(2)
print(linked_list.to_list())  # [0, 1, 3, 2]
print(linked_list.size())  # 4
print(linked_list.index_of(3))  # 2
print(linked_list.is_empty())  # False

File: Linked Lists/My Own Linked List/myOwnLinkedList.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor(value) {
    this.head = {
      value: value,
      next: null,
    };
    this.tail = this.head;
    this.length = 1;
  }
  append(value) {
    const newNode = new Node(value);
    this.tail.next = newNode;
    this.tail = newNode;
    this.length++;
    return this;
  }
  prepend(value) {
    const newNode = new Node(value);
    newNode.next = this.head;
    this.head = newNode;
    this.length++;
    return this;
  }
  printList() {
    const array = [];
    let currentNode = this.head;
    while (currentNode !== null) {
      array.push(currentNode.value);
      currentNode = currentNode.next;
    }
    console.log(array, this.length);
    return array;
  }
  printAtIndex(index) {
    let currentNode = this.head;
    for (let i = 0; i < index; i++) {
      if (!currentNode.next) {
        console.log(currentNode.value, i);
        return currentNode;
      }
      currentNode = currentNode.next;
    }
    console.log(currentNode.value, index);
    return currentNode;
  }
  insert(index, value) {
    if (index === 0) {
      this.prepend(value);
    } else if (index >= this.length) {
      this.append(value);
    } else {
      let currentNode = this.head;
      let newNode = new Node(value);
      // Get to the index
      for (let i = 1; i < index; i++) {
        currentNode = currentNode.next;
      }
      // Create a pointer to the next value
      const next = currentNode.next;
      // Break the chain: insert the new node
      currentNode.next = newNode;
      // Glue the chain: point the inserted node to saved next node
      newNode.next = next;
      this.length++;
    }
  }
  remove(index) {
    if (index === 0) {
      this.head = this.head.next;
      this.length--;
      return this.head;
    } else if (index >= this.length) {
      console.log("Nothing to remove");
      return null;
    } else {
      let currentNode = this.head;
      // Get to the index
      for (let i = 1; i < index; i++) {
        currentNode = currentNode.next;
      }
      const removedNode = currentNode.next;
      currentNode.next = currentNode.next.next;
      this.length--;
      return removedNode;
    }
  }
}

let LinkList = new LinkedList("initial value");
LinkList.append("First added Item");
LinkList.append("Second added Item");
LinkList.append("Third added Item");
LinkList.prepend("Prepended Item");
LinkList.insert(3, "Insert");
LinkList.insert(1, "Insert One");
LinkList.insert(0, "Insert Zero");
LinkList.printList();
LinkList.insert(99, "Insert Far Above");
LinkList.printList();
LinkList.remove(3);
LinkList.printList();

for (let i = 0; i < LinkList.length; i++) {
  LinkList.printAtIndex(i);
}
